src/Storages/Hive/StorageHive.cpp

```diff
@@ -0,0 +1,668 @@
+    return std::string{&path[basename_start + 1], path.length() - basename_start - 1};
+}
+
+class HiveSource : public SourceWithProgress, WithContext
+{
+public:
+    struct SourcesInfo
+    {
+        HiveMetastoreClientPtr hive_metastore_client;
+        std::string database;
+        std::string table_name;
+        HiveFiles hive_files;
+        NamesAndTypesList partition_name_types;
+
+        std::atomic<size_t> next_uri_to_read = 0;
+
+        bool need_path_column = false;
+        bool need_file_column = false;
+    };
+
+    using SourcesInfoPtr = std::shared_ptr<SourcesInfo>;
+
+    static Block getHeader(Block header, const SourcesInfoPtr & source_info)
+    {
+        if (source_info->need_path_column)
+            header.insert({DataTypeString().createColumn(), std::make_shared<DataTypeString>(), "_path"});
+        if (source_info->need_file_column)
+            header.insert({DataTypeString().createColumn(), std::make_shared<DataTypeString>(), "_file"});
+
+        return header;
+    }
+
+    static ColumnsDescription getColumnsDescription(Block header, const SourcesInfoPtr & source_info)
+    {
+        ColumnsDescription columns_description{header.getNamesAndTypesList()};
+        if (source_info->need_path_column)
+            columns_description.add({"_path", std::make_shared<DataTypeString>()});
+        if (source_info->need_file_column)
+            columns_description.add({"_file", std::make_shared<DataTypeString>()});
+        return columns_description;
+    }
+
+    HiveSource(
+        SourcesInfoPtr source_info_,
+        String hdfs_namenode_url_,
+        String format_,
+        String compression_method_,
+        Block sample_block_,
+        ContextPtr context_,
+        UInt64 max_block_size_,
+        const Names & text_input_field_names_ = {})
+        : SourceWithProgress(getHeader(sample_block_, source_info_))
+        , WithContext(context_)
+        , source_info(std::move(source_info_))
+        , hdfs_namenode_url(hdfs_namenode_url_)
+        , format(std::move(format_))
+        , compression_method(compression_method_)
+        , max_block_size(max_block_size_)
+        , sample_block(std::move(sample_block_))
+        , to_read_block(sample_block)
+        , columns_description(getColumnsDescription(sample_block, source_info))
+        , text_input_field_names(text_input_field_names_)
+        , format_settings(getFormatSettings(getContext()))
+    {
+        /// Initialize to_read_block, which is used to read data from HDFS.
+        to_read_block = sample_block;
+        for (const auto & name_type : source_info->partition_name_types)
+        {
+            to_read_block.erase(name_type.name);
+        }
+
+        /// Initialize format settings
+        format_settings.csv.delimiter = '\x01';
+        format_settings.csv.input_field_names = text_input_field_names;
+        format_settings.csv.read_bool_as_uint8 = true;
+        format_settings.defaults_for_omitted_fields = true;
```

Comment:
This setting is conrolled by user, why we always set it to true?

Reply:
When we are querying hive table, it should be true, because values of  newly added column in hive table in old files  should be default value.

Reply:
> If we defined default expression for the new column, then default values follows default expression, otherwise is zero value of its type.

Why don't we want to allow users to use setting `defaults_for_omitted_fields` to ignore default expressions and fill missed dolumns always with default expressions, not with zero values of its type?

Reply:
that is not the behavior hive, but can be viewed as a feature of clickhouse hive engine.

Reply:
Ok, let's keep it.

Reply:
The hard codes is removed.